src/lib/journey/stages/username-password.svelte

```diff
@@ -26,22 +55,26 @@
   <T key="loginHeader" />
 </h1>
 
-{#if !step}
-  <div class="tw_text-center tw_w-full tw_py-4">
-    <Spinner colorClass="tw_text-primary-light" layoutClasses="tw_h-28 tw_w-28" />
-  </div>
-{:else if step.type === 'Step'}
-  <Form bind:formEl onSubmitWhenValid={submitForm}>
+<Form bind:formEl onSubmitWhenValid={submitForm}>
+  {#if !step}
+    <div class="tw_text-center tw_w-full tw_py-4">
+      <Spinner colorClass="tw_text-primary-light" layoutClasses="tw_h-28 tw_w-28" />
+    </div>
+  {:else if step.type === 'Step'}
+    {#if failureMessage}
+      <Alert type="error" needsFocus={alertNeedsFocus}>{interpolate(failureMessageKey, null, failureMessage)}</Alert>
+    {/if}
     {#each step?.callbacks as callback, idx}
-      <svelte:component this={mapCallbackToComponent(callback)} {callback} {idx} />
+      {@const firstInvalidInput = checkValidation(callback)}
+      <svelte:component this={mapCallbackToComponent(callback)} {callback} {idx} {firstInvalidInput} />
     {/each}
     <Button width="full" style="primary" type="submit">
       <T key="loginButton" />
     </Button>
-  </Form>
-  <p class="tw_text-center tw_py-4 tw_text-secondary-dark dark:tw_text-secondary-light">
-    <T key="dontHaveAnAccount" html={true} />
-  </p>
-{:else if step.type === 'LoginSuccess'}
-  <T key="loginSucess" />
-{/if}
+    <p class="tw_text-center tw_py-4 tw_text-secondary-dark dark:tw_text-secondary-light">
+      <T key="dontHaveAnAccount" html={true} />
```

Comment:
Just a thought about this link.. It seems to make sense if you are using the widget inline, redirecting to another version of the inline widget but presenting a registration journey. However when using the modal, having the link take you to another page which may or may not have the widget in a modal makes less sense. Perhaps this should switch the widget to the registration journey if it is in a modal, instead of redirecting?

Reply:
This is all spot on. Currently, the links presented here don't do anything. There's a story that will address this, and it's found here: https://bugster.forgerock.org/jira/browse/SDKS-2053. We will end up with 4 distinct flows that will be internally managed by the widget:

1. Login
2. Registration
3. Forgot Password
4. Forgot Username

I just haven't accounted for the UX around these flows yet.

Reply:
Excellent, thanks